proposals/0003-external-calls.md

```diff
@@ -72,165 +83,195 @@ HoneyPot :: caller <- (any) {
-}
-try! alpha!.doesNothingWithArgs(x, y, z)
-// Catch can not be provided by using try!, then if the call fails then transaction reverts.
+### Code example
 
-try! alpha!.withdraw() // This flags an error as the return value is not dealt with
+The function interface of the external contract has to be specified using an "external" trait. External traits are similar to contract traits, but have a number of limitations, due to the nature of the low-level ABI of Solidity and the fact that Flint-specific features cannot be supported on Solidity contracts:
 
-try! boundReturn <- alpha!.withdraw() then {
-  // Optionally does something with boundReturn
-}
+ - Type states cannot be specified
+ - Caller protection blocks cannot be specified
+ - `mutating` or `public` keywords cannot be specified on functions
+ - Default implementations cannot be specified
+ - `Self` cannot be used
 
-// Setting hyper parameters //
-// Asset types are atomically transferred to preserve special properties.
-alpha!.transfer(Wei(200))
-alpha!.allocate(Gas(2000))
-alpha.trust() // Removes the need for a bang
-try! alpha.expensiveFunction()
-```
+Some additional caveats:
 
-### Trusted Calls
+ - Function arguments can be given labels, but these are for internal use only (since they do not affect the ABI signature)
+ - Functions can be given return types, but there is no trivial way to check if a returned value is of the required type (e.g. a `Bool` `true` value has the same representation as a `Int` `1` in the Solidity ABI)
+ - External traits have an implicit constructor, so that an address can be "cast" into the trait, allowing function calls
+ - Functions of external trait instances cannot be called using the regular function call syntax, but must use the `call` keyword, which also allows hyper-parameters to be specified
 
-#### Flint Package Manager
 ```swift
-// Creates a contract from the data stored in Flint Package Manager
-import flint://0x000... as ERCToken
-
-ERCToken.transfer(...)
+external trait Alpha {
+  func simpleFunction()
+  func functionWithArguments(value: Int, tax: Int)
+  func functionWithReturn() -> Int
+  
+  @payable
+  func expensiveFunction()
+}
 ```
-#### Source Code
-A contract's source code can be imported by:
-- Directly downloading its source files
-- Providing a web URL
-- Finding the Flint contract in the (future) Flint Package Manager
 
-```swift
-import https://flint.org/examples/contract.flint as URLContract
-import FileContract
-import Directory.Contract
+The trait can then be used in Flint code:
 
-let contract: Contract<URLContract> = deploy(URLContract)
-contract.argumentName() // Value and Gas are automatically set based upon properties
+```swift
+contract Beta {
+  func useAlpha() {
+    let someAddress: Address = 0x... // deployed Alpha contract
+    let alpha: Alpha = Alpha(address: someAddress) // using the implicit constructor
+    
+    do {
+      try call alpha.simpleFunction() // with default gas allocation: 2300
```

Comment:
Can we have `call` act like `try call`? I don't think there is value in having two keywords that only work together. 
So `call` has `try` semantics.

Reply:
Good point, I think that makes sense.